core/src/main/java/fr/sncf/osrd/conflicts/SpacingResourceGenerator.kt

```diff
@@ -172,84 +170,127 @@ class SpacingRequirementAutomaton(
     }
 
+    // Returns true if the zone is required to be clear for the signal to not be constraining.
+    // Returns false if it's either not required, or we need a longer block path to tell.
+    private fun isZoneIndexRequiredForSignal(
+        probedZoneIndex: Int,
+        pathSignal: PathSignal
+    ): Boolean {
+        val firstBlockIndex = pathSignal.minBlockPathIndex
+
+        // List of blocks to include in the simulator call
+        val blocks = mutableStaticIdxArrayListOf(incrementalPath.getBlock(firstBlockIndex))
+
+        // Index of the first zone included in the block path,
+        // used to properly index the zone state array
+        val firstSimulatedZone = incrementalPath.getBlockStartZone(firstBlockIndex)
+
+        // Number of zones included in the block path
+        var nSimulatedZones = blockInfra.getBlockPath(blocks[0]).size
+
+        // Add blocks in the block path until the probed zone is covered
+        while (probedZoneIndex - firstSimulatedZone + 1 > nSimulatedZones) {
+            if (firstBlockIndex + blocks.size >= incrementalPath.endBlockIndex) {
+                // exiting, the end of the block path has been reached
+                return false
+            }
+            val newBlock = incrementalPath.getBlock(firstBlockIndex + blocks.size)
+            blocks.add(newBlock)
+            nSimulatedZones += blockInfra.getBlockPath(newBlock).size
+        }
+
+        val zoneStates = MutableList(nSimulatedZones) { ZoneStatus.CLEAR }
+        zoneStates[probedZoneIndex - firstSimulatedZone] = ZoneStatus.OCCUPIED
+        val simulatedSignalStates =
+            simulator.evaluate(
+                rawInfra,
+                loadedSignalInfra,
+                blockInfra,
+                blocks,
+                blocks.size,
+                zoneStates,
+                ZoneStatus.OCCUPIED
+            )
+        val signalState = simulatedSignalStates[pathSignal.signal]!!
+
+        // FIXME: Have a better way to check if the signal is constraining
+        return signalState.getEnum("aspect") != "VL"
+    }
+
+    // Returns the index of the first zone that isn't required for the given signal,
+    // or null if we need more path to determine it
+    private fun findFirstNonRequiredZoneIndex(pathSignal: PathSignal): Int? {
+        // We are looking for the index `i` where `isZoneIndexRequiredForSignal` returns
+        // true at `i-1` and false at `i`. We could just iterate starting at 0, but
+        // because `i` is not that small (20 on average) and the signaling
+        // simulation calls are expensive, we prefer an approach similar to
+        // a binary-search
+
+        // The values here are determined empirically on imported infrastructures,
+        // the solutions mostly follow a gaussian distribution centered
+        // on start+20 that rarely exceeds start+40.
+        // We run a binary search on that range, and iterate one by one when the solution is above.
```

Comment:
Does this really yield any performance benefit at all? What if the bound exceeds 40?

Reply:
Roughly -30% execution time, for an arbitrary request on France repeated 15 times on each version. We go from 20 to 5 simulator calls per signal on average.

I do have some plans to improve it further, this loop is one of the main execution time bottlenecks. But this PR is large enough as it is.